test/gtest/ucp/test_ucp_proto_mock.cc

```diff
@@ -0,0 +1,408 @@
+        }
+
+        auto it = self->m_iface_attrs_funcs.find(self->m_iface_names[base]);
+        if (it != self->m_iface_attrs_funcs.end()) {
+            (it->second)(*iface_attr);
+        }
+
+        return status;
+    }
+
+    uct_component_h                                     m_component;
+    ucs::mock                                          &m_mock;
+    std::unordered_map<uct_base_iface_t *, std::string> m_iface_names;
+    std::unordered_map<std::string, iface_attr_func_t>  m_iface_attrs_funcs;
+    std::unordered_map<std::string, perf_attr_func_t>   m_perf_attrs_funcs;
+};
+
+mock_component &mock_component_test::setup_mock(const std::string &name)
+{
+    uct_component_h component = find_uct_component(name);
+    if (nullptr == component) {
+        UCS_TEST_SKIP_R("Component " + name + " not found");
+    }
+
+    mock_component *mock_comp = find(component);
+    if (nullptr == mock_comp) {
+        mock_comp = new mock_component(component, mock());
+        map().emplace(component, std::unique_ptr<mock_component>(mock_comp));
+    }
+
+    return *mock_comp;
+}
+
+struct proto_select_data {
+    std::string range_start;
+    std::string range_end;
+    std::string desc;
+    std::string config;
+};
+
+using proto_select_data_vec_t = std::vector<proto_select_data>;
+
+class test_ucp_proto_mock : public ucp_test, public mock_component_test {
+public:
+    static void get_test_variants(std::vector<ucp_test_variant> &variants)
+    {
+        add_variant(variants, UCP_FEATURE_TAG | UCP_FEATURE_AM);
+    }
+
+    virtual void init() override
+    {
+        modify_config("PROTO_ENABLE", "y");
+        modify_config("MAX_RNDV_LANES", "1");
+
+        /* Reset topo provider to force reload from config */
+        ucs_sys_topo_provider = NULL;
+        modify_config("TOPO_PRIO", "default");
+    }
+
+    virtual void cleanup() override
+    {
+        mock_component_test::cleanup();
+        ucp_test::cleanup();
+        /* Reset topo provider to not affect subsequent tests */
+        ucs_sys_topo_provider = NULL;
+    }
+
+    static void check_ep_config(entity &e, const proto_select_data_vec_t &data,
+                                ucp_proto_select_key_t key)
+    {
+        ucp_ep_config_t *config = ucp_worker_ep_config(e.worker(), 0);
+        check_proto_select(e.worker(), config->proto_select, data, key);
+    }
+
+    static void check_rkey_config(entity &e,
+                                  const proto_select_data_vec_t &data,
+                                  ucp_proto_select_key_t key)
+    {
+        ucp_rkey_config_t *config = &e.worker()->rkey_config[0];
+        check_proto_select(e.worker(), config->proto_select, data, key);
+    }
+
+    static ucp_proto_select_key_t any_key()
+    {
+        ucp_proto_select_key_t key;
+        key.u64 = UINT64_MAX;
+        return key;
+    }
+
+    void connect()
+    {
+        sender().connect(&receiver(), get_ep_params());
+        for (int i = 0; i < 10; ++i) {
```

Comment:
why needed and why 10 is enough?

Reply:
Right, it's ugly, better wait until certain condition happens, will reimplement this